### PR TITLE
Upload using pref as the recommended way

### LINKED ISSUE #5196: Conan upload receiving a pref instead of -p and from file
- The same we did with `conan get`, the pattern could process automatically the package to be uploaded, the `-p` usage could be warned as deprecated.

- Additionally, following the syntax used by other programs like `gcc`, `conan upload @filepath` could read the reference from a file.

### PR DESCRIPTION
**Pull request by lasote**
Changelog: Feature: The `conan upload` command can receive now the full package reference to upload a binary package. The `-p` argument is now deprecated.
Docs: https://github.com/conan-io/docs/pull/1300

Closes #5196

### CODE REVIEW

File: conans/client/command.py [hunk 1]
```diff
@@ -1222,7 +1228,24 @@ def upload(self, *args):
...
-        if args.query and args.package:
+        try:
+            pref = PackageReference.loads(args.pattern_or_reference, validate=True)
+            reference = pref.ref.full_repr()
+            package_id = pref.id
+        except ConanException:
+            reference = args.pattern_or_reference
+            package_id = args.package
+
+            if package_id:
```
Comment (uilianries) [inline conans/client/command.py]: I think it's not aligned with line 1243

Comment (lasote) [reply conans/client/command.py]: The `else` correspond to `try`, `except`.

File: conans/client/command.py [hunk 2]
```diff
@@ -1222,7 +1228,24 @@ def upload(self, *args):
...
+                self._user_io.out.warn("Usage of `--package` argument is deprecated."
+                                       " Use a full reference instead: "
+                                       "`conan upload [...] {}:{}`".format(reference, package_id))
+        else:
+            if args.package:
+                raise ConanException("Use a full package reference (preferred) or the `--package`"
```
Comment (uilianries) [inline conans/client/command.py]: I can't see a test case for this situation

File: conans/client/command.py [hunk 3]
```diff
@@ -91,7 +92,11 @@ def _fill_text(self, text, width, indent):
 _KEEP_SOURCE_HELP = ("Do not remove the source folder in local cache, even if the recipe changed. "
                      "Use this for testing purposes only")
 _PATTERN_OR_REFERENCE_HELP = ("Pattern or package recipe reference, e.g., '%s', "
-                              "'%s'" % (_REFERENCE_EXAMPLE, _PATTERN_EXAMPLE))
+                              "'%s'" % (_PATTERN_EXAMPLE, _REFERENCE_EXAMPLE))
+_PATTERN_REF_OR_PREF_HELP = ("Pattern, recipe reference or package reference e.g., '%s', "
+                              "'%s', '%s'" % (_PATTERN_EXAMPLE, _REFERENCE_EXAMPLE, _PREF_EXAMPLE))
+_PREF_OR_PREF_HELP = ("Recipe reference or package reference e.g., '%s', "
```
Comment (jgsogo) [inline conans/client/command.py]:
```suggestion
_REF_OR_PREF_HELP = ("Recipe reference or package reference e.g., '%s', "
```

or `_REFERENCE_OR_PREF_HELP`

File: conans/client/command.py [hunk 4]
```diff
@@ -1222,7 +1228,24 @@ def upload(self, *args):
 
         args = parser.parse_args(*args)
 
-        if args.query and args.package:
+        try:
```
Comment (jgsogo) [inline conans/client/command.py]: This `try/except/else` is exactly the same code as the one in `def get(...)` function, do you think it is worth putting it all in a function and call it from both places?

Comment (lasote) [reply conans/client/command.py]: Probably, but they are not exactly. It would feel a bit obscure I think

Comment (jgsogo) [reply conans/client/command.py]: Ok, it was just a suggestion 👍

**Comment by jgsogo** [comment]
Update the docs and it's done!

**Comment by lasote** [comment]
Docs added! Push the button!

**Comment by jgsogo** [comment]
shakira

**Pull request merged**